Initial implementation of 3Di encoding processing using a Python library (via venv)

vrbouza opened the pull request:
This PR adds the feature of processing directly PDB files with protein information with sketchlib by

1. Parse the PDB files.
2. Encode the information using the 3Di alphabet (from [Foldseek](https://github.com/steineggerlab/foldseek) ).
3. Use aaHash as rolling hashing function and do standard sketching.

This implementation has some restrictions that should be made in the future configurable through parameters or lifted: any protein must have at least 100 aa and each of its chains, 64 aa. In addition, if the encoding fails with a chain, it is completely discarded.

To parse the pdb files and do the encoding we use PDBParser and the [mimi3di](https://github.com/althonos/mini3di) library. This is not ideal, as it requires to create a venv and use python alongside rust to run it.

The encoding is an extra feature that must be specified to be compiled (-F 3di).

Resolves #20 .

johnlees commented on Cargo.toml:
```diff
@@ -45,6 +54,9 @@ anyhow = "1.0"
 num_cpus = "1.0"
 
+[build-dependencies]
+pyo3-build-config = {version = "*", optional = true}
```
Maybe we should specify major version?

vrbouza replied:
Yes, done (just set the current one)

johnlees commented on src/hashing/aahash_iterator.rs:
```diff
@@ -107,6 +111,16 @@ impl AaHashIterator {
     }
 
+    pub fn from_3di_file(file: &str, concat_fasta: bool) -> Vec<Self> {
+        Self::new(file, AaLevel::Level1, concat_fasta)
```
`concat_fasta` I think would always be false for a PDB file?

vrbouza replied:
Yes, modified this, but waiting for confirmation on the other comment

johnlees commented on src/sketch.rs:
```diff
@@ -250,26 +267,64 @@ pub fn sketch_files(
+
+
+                    // Temporal (?) restrictions to the size of proteins and chains
+                    if *seq_type == HashType::PDB {
```
Could these checks go in `structures.rs` and be called before `from_3di_string` and `from_3di_file`?

vrbouza replied:
Yes... but you'll still need to modify something here (in sketch.rs) IF you still want to panic whenever you find a sequence with 0. The thing is that in this big iterate, you go over input_files: if you filter things in structures.rs, you might (you do) get a smaller struct_strings vector, and thus there is issue because you are expecting the same size as in input_files (as I was developing, I decided to not change the code so much and filter afterwards, despite being less efficient).

If we move the filtering to structures, we need to remove the panic somehow when treating PDB files (e.g. filtering those empty "sequences" in those cases), do you see it fit, or maybe you prefer other way? (We could also change more the iterator etc., but I'm not sure we want that...).

PS: I also discovered that the levels of aaHash were fixed to level1 with the PDB files: just removed them in case we want to explore there as we spoke in our last chat (always assuming the 3Di dict. keeps close the same symbols as those of the nucleotides)

johnlees replied:
> PS: I also discovered that the levels of aaHash were fixed to level1 with the PDB files: just removed them in case we want to explore there as we spoke in our last chat (always assuming the 3Di dict. keeps close the same symbols as those of the nucleotides)

From memory I don't think the symbols correspond to similar motifs in the same way as AAs, but we should double-check in the foldseek paper

johnlees replied:
Rather than filtering to a lower number of sequences, I think desired behaviour might be just to panic if anything fails and then the user can check, filter and rerun themselves (this is what worked best in PopPUNK)

vrbouza replied:
Ok, so:
- I checked Foldseek's paper/docs and they do not mention anything regarding similarity between symbols as with aminoacids, so I reverted the possibility to use different levels of aaHash.
- I removed the filters that I added, now I filter the pdb files before.
- Also removed the `concat_fasta` thing from previous comment.

johnlees reviewed (changes requested):
Thanks! A few suggestions in review.
Is it possible to update the .github/actions to run on your fork? Or should we make all changes from a branch in the main repo in future?

vrbouza commented:
> Thanks! A few suggestions in review. Is it possible to update the .github/actions to run on your fork? Or should we make all changes from a branch in the main repo in future?

The actions run in my repo & branches! (Except the version one). I can move to do development in a branch in the bacpop repo if you prefer, I don't have a preference. In any case, perhaps it would be useful to have a cargo build -F 3di also somewhere in the CI...

johnlees commented:
> > Thanks! A few suggestions in review. Is it possible to update the .github/actions to run on your fork? Or should we make all changes from a branch in the main repo in future?
> 
> The actions run in my repo & branches! (Except the version one). I can move to do development in a branch in the bacpop repo if you prefer, I don't have a preference. 

Yeah maybe let's do that for the next ones (i.e. dev branch in bacpop repo) as for whatever reason I don't see the actions in the PR here.

> In any case, perhaps it would be useful to have a cargo build -F 3di also somewhere in the CI...

Good idea! We could add this in this PR?

vrbouza commented:
> Yeah maybe let's do that for the next ones (i.e. dev branch in bacpop repo) as for whatever reason I don't see the actions in the PR here.
Will do! Anyway for this PR I am checking that they run correctly before commenting in the PR.

> > In any case, perhaps it would be useful to have a cargo build -F 3di also somewhere in the CI...
> Good idea! We could add this in this PR?
Added in the last commit, just modified the CI and added a new "run" task in the ci.yaml Actions workflow.

PS: a nice TODO for the future would be to parallelise this encoding. In python alone I have it already, and while I was waiting for my checks to run and confirm that it still works I'm half-way into implementing it, but perhaps the moment I try the parallelisation with the python-rust interface will be tricky...

johnlees reviewed (approved):
Ok, nice. Let's raise parallelising the conversion as an issue

johnlees commented:
@vrbouza sorry did I mean to merge these already? Looks like it's ready?

vrbouza commented:
I think so! I thought it was already merged and forgot completely about this pr hehe

johnlees commented:
Yeah me too! Merged now
